Add image schema to schema data

KiOui opened the pull request:
## Context
<!--
What do we want to achieve with this PR? Why did we write this code?
-->

* We want to add image schema data to the schema we are currently rendering.

When merging this PR, [this PR for Yoast SEO WooCommerce](https://github.com/Yoast/wpseo-woocommerce/pull/833) should also be merged.

## Summary

<!--
Attach one of the following labels to the PR: `changelog: bugfix`, `changelog: enhancement`, `changelog: other`, `changelog: non-user-facing`.
If the changelog item is a bugfix, please use the following sentence structure: Fixes a bug where ... would ... (when ...).
If the changelog item is meant for the changelog of another add-on, start your changelog item with the name of that add-on's repo between square brackets, for example: * [wordpress-seo-premium] Fixes a bug where ....
If the changelog items is meant for the changelog of a javascript package, specify between square brackets in which package changelog the item should be included, for example: * [@yoast/components] Fixes a bug where ....
If the same changelog item is applicable to multiple changelogs/add-ons, add a separate changelog item for all of them.
-->
This PR can be summarized in the following changelog entry:

* Adds multiple imageObject pieces in the schema, for every image found in the page content.

## Relevant technical choices:

* The featured image was previously rendered with a custom tag. This now uses a normal schema id.
* Refactored `WPSEO_Sitemap_Image_Parser` such that we are able to use the image parsing code in other places as well.
* Removed schema code deprecated since 14.0 because it also used the `PRIMARY_IMAGE_HASH` that we are not using in this PR anymore. The [LSX Tour Operator](https://wordpress.org/plugins/tour-operator/) plugin that was using this and we decided to still remove this because of how long it was deprecated for.
* We have added a class_alias for the Main_Images class because there are plugins who rely on this specific classname.

## Test instructions
<!--
Please follow these guidelines when creating test instructions:
- Please provide step-by-step instructions how to reproduce the issue, if applicable.
- Write step-by-step test instructions aimed at non-tech-savvy users, even if the PR is not user-facing.
-->
### Test instructions for the acceptance test before the PR gets merged
This PR can be acceptance tested by following these steps:

Check whether image schema data works for posts without images:
* Make sure that no organisation or person logo is set in the Yoast SEO settings (it might be good to start with a clean site).
* Go to the post editor and create a new empty post.
* Save the post and go to the public (non-editor) view of the post you just created.
* Go to the HTML source of the post and check that there is no `primaryImageOfPage` entry in the `WebPage` object. Also check that there is no `image` entry in the `WebPage` object and that there are no `ImageObject` objects within the `@graph` list (unless it's a post, in which case there should be just one `ImageObject` object, the one in the `image` property of the `person` object).

Check whether image schema data works for a post with only an image:
* Go to the post editor and create a new empty post. Add an image to the post.
* Save the post and go to the public (non-editor) view of the post you just created.
* Go to the HTML source of the post and check that there is exactly one `ImageObject` object within the `@graph` list (the ID for the image should be the URL). Also check that the `primaryImageOfPage` entry in the `WebPage` object is set to the same ID and that the `image` entry in the `WebPage` object also includes the same ID as the only image.

Check whether image schema data works for a post with only a primary image: 
* Go to the post editor and create a new empty post. Set a primary image for the post.
* Save the post and go to the public (non-editor) view of the post you just created.
* Go to the HTML source of the post and check that there is exactly one `ImageObject` object within the `@graph` list (the ID for the image should be the URL). Also check that the `primaryImageOfPage` entry in the `WebPage` object is set to the same ID and that the `image` entry in the `WebPage` object also includes the same ID as the only image.

Check whether post content images are listed within the schema data:
* Go to the post editor and create a new empty post (without a featured image). Add a couple of images to the post. Also add some duplicate images (images that are the same).
* Save the post and go to the public (non-editor) view of the post you just created.
* Go to the HTML source of the post and check that there is exactly one `ImageObject` object for each *unique* image added to the post in the `@graph` list (the ID for the image should be the URL). Also check that the `primaryImageOfPage` entry in the `WebPage` object is *not* set and that the `image` entry in the `WebPage` object also includes the ID's of the images that are within the post content.

Check whether post content images and primary image do not duplicate within the schema data:
* Go to the post editor and create a new empty post. Add a featured image and add the same image within the post content.
* Save the post and go to the public (non-editor) view of the post you just created.
* Go to the HTML source of the post and check that there is exactly one `ImageObject` object for the image added to the post in the `@graph` list (the ID for the image should be the URL). Also check that the `primaryImageOfPage` entry in the `WebPage` object is set to the same ID and that the `image` entry in the `WebPage` object also includes the ID of the one image added.

Check whether post content images (external) are listed within the schema data:
* Go to the post editor and create a new empty post. Add an external image to that post via the URL (such as `https://helpx.adobe.com/content/dam/help/en/photoshop/using/convert-color-image-black-white/jcr_content/main-pars/before_and_after/image-after/Landscape-BW.jpg?size=large`).
* Save the post and go to the public (non-editor) view of the post you just created.
* Go to the HTML source of the post and check that there is an `ImageObject` object for the external image added to the post in the `@graph` list (the ID for the image should be the URL). Also check that the `image` entry in the `WebPage` object includes the ID of the externally added image.
* Also make sure the entire url is present in the @id and @contentUrl (so with the ?size=large part)

Check whether facebook/twitter image is listed within the schema data:
* Go to the post editor and create a new empty post. Add an image under the Yoast settings -> Social -> Facebook image.
* Save the post and go to the public (non-editor) view of the post you just created.
* Go to the HTML source of the post and check that there is an `ImageObject` object for the facebook image added. Also check that the `image` entry in the `WebPage` object also includes the same ID as the Facebook image.
* Test the same for the twitter image.

Check whether the facebook/twitter image and content images do not duplicate:
* Go to the post editor and create a new empty post. Add an image under the Yoast settings -> Social -> Facebook image. Also add images within the page content and a features image (add the same image as the featured image and the same in the page content).
* Save the post and go to the public (non-editor) view of the post you just created.
* Go to the HTML source of the post and check that there are no duplicated images as `ImageObject`s.
* Test the same for the twitter image.

Check whether the Open Graph images are displayed correctly (e.g. with https when the site is using https).
* Setup the test site with https (for `basic.wordpress.test` this is done by default).
* Make sure that the WordPress address URL and the Site Address on [https://basic.wordpress.test/wp-admin/options-general.php](https://basic.wordpress.test/wp-admin/options-general.php) are set to use `http` in the settings.
* Go to the post editor and create a new empty post. Add an image under the Yoast settings -> Social -> Facebook image. 
* Save the post and go to the public (non-editor) view of the post you just created.
* Go to the HTML source of the post and check that the `ImageObject` in the `@graph` is displaying a `url` with `https`. Also check that the `@id` corresponds to the `url` and that the `@id` within the `image` block of `WebPage` has `https`.

Check whether resized images are displayed correctly.
* Go to the post editor and create a new empty post. Add an image in the post content.
* In the right sidebar of the image, under Image size, select something else than `Full Size`.
* Save the post and go to the public (non-editor) view of the post you just created.
* Check whether the image URL in the `ImageObject` is correctly displayed as the `@id` and `url` (with the image size) and check that the `image` array under the `WebPage` also has the correct URL (with the image size).

#### Relevant test scenarios
* [ ] Changes should be tested with the browser console open
* [x] Changes should be tested on different posts/pages/taxonomies/custom post types/custom taxonomies
* [x] Changes should be tested on different editors (Block/Classic/Elementor/other)
* [ ] Changes should be tested on different browsers
* [ ] Changes should be tested on multisite
<!--
If you have checked any of the above cases, please add some context about the reason, what to check in the console,
which type/editor/browser should be tested in particular, multisite with subfolders or subdomains, etc.
-->

### Test instructions for QA when the code is in the RC
<!--
Sometimes some steps from the test instructions for the acceptance test aren't relevant anymore once the code has been merged or the feature is complete. If that is the case, do not check the checkbox below.
QA is our Quality Assurance team. The RC is the release candidate zip that is tested before a release
-->

* [x] QA should use the same steps as above.

<!--
If the above checkbox has not been checked, write down all steps QA should take to test this PR, not only the difference with the acceptance test steps. If QA should use the test instructions specified on the epic, paste a link to the relevant comment on the epic.
-->
QA can test this PR by following these steps:

## Impact check
<!--
Sometimes PRs have a bigger impact than is suggested in the user-facing changes. In such cases,
additional (regression) testing might be necessary. To make it clear what parts might need additional testing, please outline which parts of the plugin have been impacted by this PR.
-->
This PR affects the following parts of the plugin, which may require extra testing:

* This PR affects the helper that generates the image data for the sitemap files - for example, try a post with an image, a resized image and a feature image and verify that the sitemap with this RC/PR and without it, they show the same number of images in the sitemap.
* This PR affects the helper that helps generate the image objects in the schema data. This might have an impact on Yoast's other plugins that use certain image IDs for linking images (see the thread below for more information).

## UI changes

* [ ] This PR changes the UI in the plugin. I have added the 'UI change' label to this PR.

## Other environments

* [ ] This PR also affects Shopify. I have added a changelog entry starting with `[shopify-seo]`, added test instructions for Shopify and attached the `Shopify` label to this PR.

## Documentation

* [ ] I have written documentation for this change.

## Quality assurance

* [x] I have tested this code to the best of my abilities
* [ ] I have added unit tests to verify the code works as intended
* [ ] If any part of the code is behind a feature flag, my test instructions also cover cases where the feature flag is switched off.
* [ ] I have written this PR in accordance with my team's definition of done.

Fixes [PC-19](https://yoast.atlassian.net/browse/PC-19)
Fixes [PC-20](https://yoast.atlassian.net/browse/PC-20)

KiOui commented:
Things still left to do:

- [ ] Test the speed of the current implementation.
- [ ] Verify whether the decoupledness of the current implementation should be changed (e.g. there are somewhat the same functions in `Webpage` and `Images`.
- [ ] Fix the non-working sitemap class due to dependency injection.

thijsoo commented on src/generators/schema/images.php:
```diff
@@ -0,0 +1,152 @@
+	public function generate() {
+		$graph = [];
+
+		$this->add_primary_image( $graph );
```
I think we should not use a pass by ref here. I think it is more readable to just use $graph = $this

thijsoo commented on src/generators/schema/webpage.php:
```diff
@@ -45,7 +46,7 @@ public function generate() {
 		}
 
-		$this->add_image( $data );
+		$this->add_images( $data );
```
Same as above, on line 56 $data = $this-> is also used. I think we should do the same

thijsoo reviewed (changes requested):
I did an ACC and I found that for step ` Also check that there is no image entry in the WebPage object` I did have an image in my website schema, it was my logo, so I am not sure if this is expected behaviour.
Also I think a couple of small things need to be fixed in the code :)

enricobattocchi commented on src/surfaces/helpers-surface.php:
```diff
@@ -3,6 +3,7 @@
 namespace Yoast\WP\SEO\Surfaces;
 
 use Yoast\WP\SEO\Helpers;
+use Yoast\WP\SEO\Helpers\Image_Helper;
 use YoastSEO_Vendor\Symfony\Component\DependencyInjection\ContainerInterface;
```
I don't think we need this

```suggestion
use Yoast\WP\SEO\Helpers;
use YoastSEO_Vendor\Symfony\Component\DependencyInjection\ContainerInterface;
```

KiOui commented:
I also saw some small things that still need to be changed. In some parts of the schema we are still using the `#primaryimage` as the ID (such as in the `Article` -> `image` object). A quick search for `#primaryimage` and replacing it with the image URL should work.

KiOui commented:
Also if we want to change the `@id` of every image to the URL of that image, we should take a look at all the schema that outputs an image as the Person and Organisation are using a custom ID as well.

KiOui commented:
- Public function add_image in webpage.php has been changed to a protected function add_primary_image. We might want to add a method that replaces add_image such that this method still works.
- We have to check whether the const PRIMARY_IMAGE_HASH in schema-ids.php and class-schema-ids.php are used in any of Yoast’s plugins or third-party plugins as the variable is public.
- We have to check the following plugins for their schema generation:
    - Premium
        - Only thing to check is `src/schema-templates`.
    - WPSEO WooCommerce
        - The product image uses the #primaryimage tag, [this PR](https://github.com/Yoast/wpseo-woocommerce/pull/833) fixes that.
    - WPSEO Video
        - Is not using anything that was adjusted in Free, the plugin is using the `VIDEO_HASH` for the video identifier.
    - WPSEO Local
        - Is not using anything that was adjusted in Free, the plugin is using other hashes defined in the plugin itself for several identifiers.
    - WPSEO News
        - Is not using anything that was adjusted in Free, the plugin is using the `ORGANISATION_HASH` for the copyright holder identifier.
- We changed the `Main_Image` class and added all other images within the `Images` class in the schema as well. We should check whether this causes any issues.
- We might need to change the @ids for the Person and Organization image objects as well.

jdevalk commented on src/helpers/image-helper.php:
```diff
@@ -46,6 +50,41 @@ class Image_Helper {
+	 *
+	 * @var string
+	 */
+	protected $home_url = '';
```
It's weird that this should be in this class. Shouldn't this be taken from elsewhere, and maybe made public there?

jdevalk commented on src/helpers/image-helper.php:
```diff
@@ -46,6 +50,41 @@ class Image_Helper {
+	 *
+	 * @var string
+	 */
+	protected $host = '';
```
Same as `$home_url` above.

jdevalk commented on src/helpers/image-helper.php:
```diff
@@ -46,6 +50,41 @@ class Image_Helper {
+	 *
+	 * @var string
+	 */
+	protected $scheme = 'http';
```
Same as `$home_url` above.

jdevalk commented on src/helpers/image-helper.php:
```diff
@@ -46,6 +50,41 @@ class Image_Helper {
+	 *
+	 * @var string
+	 */
+	protected $charset = 'UTF-8';
```
Same as `$home_url` above.

jdevalk commented on src/helpers/image-helper.php:
```diff
@@ -61,14 +100,27 @@ public function __construct(
+
+		if ( ! empty( $parsed_home['scheme'] ) ) {
+			$this->scheme = $parsed_home['scheme'];
+		}
```
All of the above logic feels like it should live either in a separate helper or higher order somewhere.

jdevalk reviewed (changes requested):
I feel like we're duplicating some code that we already have somewhere, and if not, it should be extracted into a separate helper and we should scan the codebase for where else we do the same things.

KiOui commented:
The class `WPSEO_Content_Images` in `/inc/class-wpseo-content-images.php` also does some image parsing that should be refactored to the `Image_Helper` at some point.

jdevalk commented on src/helpers/image-helper.php:
```diff
@@ -46,6 +50,13 @@ class Image_Helper {
 	 */
 	private $url_helper;
 
+	/**
+	 * Holds blog charset value for use in DOM parsing.
+	 *
+	 * @var string
+	 */
+	protected $charset = 'UTF-8';
+
```
Looks like you're not using this property at all, in which case it's best to get rid of it.

KiOui replied:
Good catch! Forgot to remove it.

jdevalk reviewed (changes requested):
One more thing :)

pls78 commented on src/generators/schema/webpage.php:
```diff
@@ -96,13 +97,119 @@ public function add_author( $data, $post ) {
 	 */
-	public function add_image( &$data ) {
+	public function add_images( $data ) {
+		$data = $this->add_primary_image( $data );
```
I would move this call outside the add_images function, as it takes care of the primery image-related properties.

pls78 commented:
CR: ✅

leonidasmi commented:
There was a 
```
PHP Notice:  Trying to get property 'post_content' of non-object in /var/www/html/wp-content/plugins/wordpress-seo/src/context/meta-tags-context.php on line 228
```
notice showing on the edit pages of posts, which I took care with [this commit](https://github.com/Yoast/wordpress-seo/pull/18916/commits/d70fa65d7fdee2c0b08a5b861c2d78b8fe1b8a90).

leonidasmi commented:
Acceptance Testing: ❌ 

I found 2 issues with having resized images in the posts:

**Post with resized image in the content and no other images**
* Have a post with no feature image and add in the content of the post, an image
* Click on the image and on the sidebar, select `Thumbnail` for the Image size
* Save the post
* In the frontend, there's an `@id` in the `image` attribute of the `webpage` that doesnt reference any `imageobject` in the schema and that's the id of the resized image. Example of such a schema, [here](https://yoast.slack.com/archives/C03KU0EHCNQ/p1664539641843939?thread_ts=1664517895.576709&cid=C03KU0EHCNQ).

**Post with resized image in the content and with a feature image**
* Have a post with a feature image and add in the content of the post, an image
* Click on the image and on the sidebar, select `Thumbnail` for the Image size
* Save the post
* In the schema of the post, there's no mention of that resized image, just a mention of its full-sized counter-part. Example of such a schema, [here](https://yoast.slack.com/archives/C03KU0EHCNQ/p1664539797384959?thread_ts=1664517895.576709&cid=C03KU0EHCNQ).

KiOui commented:
> Acceptance Testing: ❌
> 
> I found 2 issues with having resized images in the posts:
> 
> **Post with resized image in the content and no other images**
> 
> * Have a post with no feature image and add in the content of the post, an image
> * Click on the image and on the sidebar, select `Thumbnail` for the Image size
> * Save the post
> * In the frontend, there's an `@id` in the `image` attribute of the `webpage` that doesnt reference any `imageobject` in the schema and that's the id of the resized image. Example of such a schema, [here](https://yoast.slack.com/archives/C03KU0EHCNQ/p1664539641843939?thread_ts=1664517895.576709&cid=C03KU0EHCNQ).
> 
> **Post with resized image in the content and with a feature image**
> 
> * Have a post with a feature image and add in the content of the post, an image
> * Click on the image and on the sidebar, select `Thumbnail` for the Image size
> * Save the post
> * In the schema of the post, there's no mention of that resized image, just a mention of its full-sized counter-part. Example of such a schema, [here](https://yoast.slack.com/archives/C03KU0EHCNQ/p1664539797384959?thread_ts=1664517895.576709&cid=C03KU0EHCNQ).

This happened due to the `Image_Helper` not taking into account the size of the image and just getting the image URL with size `full` every time. It is now fixed by extracting the image size from the URL (which we can do because WordPress makes sure that images that are resized get `-[width]x[height]` appended to their file name) and passing the image size with the `Image` value to the generators.

The change above also fixes a bug in the `open_graph_images` being added with `http` to the graph instead of `https`. We already fixed this for the `Images` class but not yet for the `WebPage` class.

leonidasmi commented:
I notice a small issue with this fix, described with these instructions:

* Have a post with no feature image and add in the content of the post, an image
* Click on the image and on the sidebar, select `Thumbnail` for the Image size
* Save the post
* In the schema of that Imageobject, the width/height attributes are showing wrong values, `640&480` instead of `300&225`: 
```
{
	"@type": "ImageObject",
	"inLanguage": "en-US",
	"@id": "http://basic.wordpress.test/wp-content/uploads/2022/10/WordPress7-2-300x225.jpg",
	"url": "http://basic.wordpress.test/wp-content/uploads/2022/10/WordPress7-2-300x225.jpg",
	"contentUrl": "http://basic.wordpress.test/wp-content/uploads/2022/10/WordPress7-2-300x225.jpg",
	"width": 640,
	"height": 480
},
```

leonidasmi commented:
I created a subtask for the above 👍

leonidasmi commented:
There's a fatal error when accessing the Sitemap:
```
Fatal error: Uncaught Error: Call to undefined method Yoast\WP\SEO\Helpers\Image_Helper::image_url()
in /var/www/html/wp-content/plugins/wordpress-seo/inc/sitemaps/class-sitemap-image-parser.php on line 166
```
and I'm gonna push a commit to fix it.

thijsoo commented:
CR'ed the fix from Leonidas for the sitemap 👍

leonidasmi commented:
There's an issue with external images: 

* when they happen to not be the primaryImageOfPage
* they have empty `url` and `contentUrl` attributes in their `ImageObject` node:
```
{
    "@type": "ImageObject",
    "inLanguage": "en-US",
    "@id": "https://helpx.adobe.com/content/dam/help/en/photoshop/using/convert-color-image-black-white/jcr_content/main-pars/before_and_after/image-after/Landscape-BW.jpg?size=large",
    "url": "",
    "contentUrl": ""
},
```

I'll create a subtask so that we can solve this in the feature branch

leonidasmi commented:
Acceptance Test is 👍 (apart from the aforementioned, non-blocking issues)
